File: vm/complete/retro.rb
#!/usr/bin/env ruby

# Ngaro VM
# Copyright (c) 2010 - 2011, Charles Childers
# Copyright (c) 2011, Foucist
# -----------------------------------------------------------------------------
# RubyFORTH -- Copyright (C) 2007-8, Marc Simpson (GPL). 
# 
# Stack module
# 
# Relicensed for Retro under the ISC License (thanks Marc!)
# -----------------------------------------------------------------------------
class Stack
  def initialize(size)
    @contents = Array.new(size)
    @offset = -1
  end
  def nth(n)
    @contents[@offset - n]
  end
  def tos
    if @offset < 0
      puts "Error: stack underflow"
      false
    else
      @contents[@offset]
    end
  end
  def push(n)
    @offset += 1
    @contents[@offset] = n
  end
  def pop
    if @offset < 0
      puts "Error: stack underflow"
      @offset = -1
      throw("toplevel")
      return
    end
    @offset -= 1
    @contents[@offset + 1]
  end
  def depth
    @offset + 1
  end
  def swap
    x = self.pop ; y = self.pop
    self.push(x) ; self.push(y)
  end
end

@ip = 0                     # instruction pointer
@stack   = Stack.new(128)   # data stack
@address = Stack.new(1024)  # return stack
@memory  = {}               # simulated ram
@ports   = {}               # io ports
@files   = {}               # file handles
@inputs  = {}               # input sources
@isp     = 0                # input source pointer

# -----------------------------------------------------------------------------
# Workaround for keyboard input to let the same source run under
# Ruby 1.8 and 1.9
#
def get_char()
  a = 0
  if @isp == 0 then
    case RUBY_VERSION.split('.')[1]
      when "9" then a = STDIN.getbyte
      when "8" then a = STDIN.getc
    end
  else
    begin
      case RUBY_VERSION.split('.')[1]
        when "9" then a = @inputs[@isp].readbyte
        when "8" then a = @inputs[@isp].readchar
      end
    rescue EOFError
      a = 0
      @inputs[@isp].close
      @isp = @isp - 1
    end
  end
  return a
end
# -----------------------------------------------------------------------------
def do_env()
  req = @stack.pop
  dest = @stack.pop
  key = ''
  while @memory[req] != 0 do
    key = key + @memory[req].chr
    req = req + 1
  end
  req = ENV[key]
  @memory[dest] = 0
  req.each_byte do |c|
    @memory[dest] = c
    dest = dest + 1
    @memory[dest] = 0
  end
  return 0
end
# -----------------------------------------------------------------------------
def handle_devices()
  if @ports[0] == 1 then return end
  if @ports[0] == 0 then
    @ports[0] = 1
    case @ports[1]
    when 1 then
      @ports[1] = get_char()
      if @ports[1] == 13 then @ports[1] = 10 end
    end
  end
  if @ports[2] == 1 then
    if @stack.tos < 0 then
      puts "\e[H\e[2J"
    elsif @stack.tos > 0 and @stack.tos < 128 then
      print @stack.tos.chr
      STDOUT.flush
    end
    @stack.pop
    @ports[2] = 0
  end

  # File I/O
  case @ports[4]
  when  1 then   # save image
    f = File.new("retroImage", "wb")
    (0..@memory[3]).each do |i|
      f.write([@memory[i]].pack("i"))
    end
    f.close
    @ports[4] = 0
  when  2 then   # include file
    name = ""
    i = @stack.pop
    while @memory[i] != 0 do
      name = name + @memory[i].chr
      i = i + 1
    end
    if File::exists?(name) then
      @isp = @isp + 1
      @inputs[@isp] = File.open(name, 'r')
    end
    @ports[4] = 0
  when -1 then   # file open
    slot = 1
    i = 1
    while i < 8 do
      if @files[i] == 0 then
        slot = i
      end
      i = i + 1
    end
    mode = @stack.pop
    name = ""
    i = @stack.pop
    while @memory[i] != 0 do
      name = name + @memory[i].chr
      i = i + 1
    end
    @files[slot] = case mode
                   when 0 then if File::exists?(name) then File.open(name, 'r') else slot = 0 end
                   when 1 then File.open(name, 'w')
                   when 2 then File.open(name, 'a')
                   when 3 then if File::exists?(name) then File.open(name, 'r+') else slot = 0 end
                   end
    @ports[4] = slot
  when -2 then   # file read
    slot = @stack.pop
    case RUBY_VERSION.split('.')[1]
      when "9" then @ports[4] = @files[slot].readbyte
      when "8" then @ports[4] = @files[slot].readchar
    end
  when -3 then   # file write
    slot = @stack.pop
    key = @stack.pop
    @files[slot].write(key.chr)
    @ports[4] = 1
  when -4 then   # file close
    slot = @stack.pop
    @files[slot].close
    @files[slot] = 0
    @ports[4] = 0
  when -5 then   # file position
    slot = @stack.pop
    @ports[4] = @files[slot].pos
  when -6 then   # file seek
    slot = @stack.pop
    offset = @stack.pop
    @files[slot].seek(offset, IO::SEEK_SET)
    @ports[4] = 0
  when -7 then   # file size
    slot = @stack.pop
    offset = @files[slot].pos
    @files[slot].seek(0, IO::SEEK_END)
    @ports[4] = @files[slot].pos
    @files[slot].seek(offset, IO::SEEK_SET)
  when -8 then   # file delete
    name = ""
    i = @stack.pop
    while @memory[i] != 0 do
      name = name + @memory[i].chr
      i = i + 1
    end
    i = 0
    if File::exists?(name) then
      File.delete(name)
      i = 1
    end
    @ports[4] = i
  end

  # Capability Queries
  @ports[5] = case @ports[5]
              when -1 then 1000000 # memory size
              when -2 then 0       # canvas exists?
              when -3 then 0       # canvas width
              when -4 then 0       # canvas height
              when -5 then @stack.depth   # stack depth
              when -6 then @address.depth # address stack depth
              when -7 then 0       # mouse exists?
              when -8 then Time.now.to_i        # time
              when -9 then (@ip = 1000000) && 0 # exit vm
              when -10 then do_env # environment
              when -11 then 0      # console width
              when -12 then 0      # console height
              when -13 then 32     # bits per cell
              when -14 then 0      # endian
              when -15 then 1      # console support
              when -16 then 128    # data stack max depth
              when -17 then 1024   # address stack max depth
              end

  case @ports[8]
    when 1 then
      @stack.swap
      print "\33[" + @stack.pop.to_s + ";" + @stack.pop.to_s + "H"
    when 2 then
      print "\33[3" + @stack.pop.to_s + "m"
    when 3 then
      print "\33[4" + @stack.pop.to_s + "m"
  end
  @ports[8] = 0


end
# -----------------------------------------------------------------------------
def process()
  case @memory[@ip] # opcode
  when 0 then 0  # nop
  when 1 then  # lit
    @ip = @ip + 1
    @stack.push(@memory[@ip])
  when 2 then  # dup
    @stack.push(@stack.tos)
  when 3 then  # drop
    @stack.pop
  when 4 then  # swap
    @stack.swap
  when 5 then  # push
    @address.push(@stack.pop)
  when 6 then  # pop
    @stack.push(@address.pop)
  when 7 then  # loop
    @stack.push(@stack.pop - 1)
    if @stack.tos != 0 and @stack.tos > -1 then
      @ip = @ip + 1
      @ip = @memory[@ip] - 1
    else
      @ip = @ip + 1
      @stack.pop
    end
  when 8 then  # jump
    @ip = @ip + 1
    @ip = @memory[@ip] - 1
    if @memory[@ip + 1] == 0 then @ip = @ip + 1 end
    if @memory[@ip + 1] == 0 then @ip = @ip + 1 end
  when 9 then  # return
    @ip = @address.pop
    if @memory[@ip + 1] == 0 then @ip = @ip + 1 end
    if @memory[@ip + 1] == 0 then @ip = @ip + 1 end
  when 10 then  # >= jump
    @ip = @ip + 1
    if @stack.nth(1) > @stack.tos then
      @ip = @memory[@ip] - 1
    end
    @stack.pop
    @stack.pop
  when 11 then  # <= jump
    @ip = @ip + 1
    if @stack.nth(1) < @stack.tos then
      @ip = @memory[@ip] - 1
    end
    @stack.pop
    @stack.pop
  when 12 then  # != jump
    @ip = @ip + 1
    if @stack.nth(1) != @stack.tos then
      @ip = @memory[@ip] - 1
    end
    @stack.pop
    @stack.pop
  when 13 then  # == jump
    @ip = @ip + 1
    if @stack.nth(1) == @stack.tos then
      @ip = @memory[@ip] - 1
    end
    @stack.pop
    @stack.pop
  when 14 then  # @
    @stack.push(@memory[@stack.pop])
  when 15 then  # !
    @memory[@stack.pop] = @stack.pop
  when 16 then  # +
    @stack.push(@stack.pop + @stack.pop)
  when 17 then  # -
    @stack.swap
    @stack.push(@stack.pop - @stack.pop)
  when 18 then  # *
    @stack.swap
    @stack.push(@stack.pop * @stack.pop)
  when 19 then  # /mod
    b = @stack.pop
    a = @stack.pop
    x = b.abs
    y = a.abs
    q = (y / x).floor
    r =  y % x
    if a < 0 and b < 0 then
      r *= -1
    elsif a > 0 and b < 0 then
      q *= -1
    elsif a < 0 and b > 0 then
      r *= -1
      q *= -1
    end
    @stack.push(r)
    @stack.push(q)
  when 20 then  # and
    a = @stack.pop
    b = @stack.pop
    @stack.push(b & a)
  when 21 then  # or
    a = @stack.pop
    b = @stack.pop
    @stack.push(b | a)
  when 22 then  # xor
    a = @stack.pop
    b = @stack.pop
    @stack.push(b ^ a)
  when 23 then  # <<
    a = @stack.pop
    b = @stack.pop
    @stack.push(b << a)
  when 24 then  # >>
    a = @stack.pop
    b = @stack.pop
    @stack.push(b >> a)
  when 25 then  # 0;
    if @stack.tos == 0 then
      @stack.pop
      @ip = @address.pop
    end
  when 26 then  # inc
    @stack.push(@stack.pop + 1)
  when 27 then  # dec
    @stack.push(@stack.pop - 1)
  when 28 then  # in
    a = @stack.pop
    @stack.push(@ports[a])
    @ports[a] = 0
  when 29 then  # out
    @ports[@stack.pop] = @stack.pop
  when 30 then  # wait
    handle_devices()
  else                     # call (implicit)
    @address.push(@ip)
    @ip = @memory[@ip] - 1
    if @memory[@ip + 1] == 0 then @ip = @ip + 1 end
    if @memory[@ip + 1] == 0 then @ip = @ip + 1 end
  end
end
# -----------------------------------------------------------------------------
@cells = (File.size("retroImage") / 4)
File.open("retroImage", "r") do |f|
  (0..(@cells-1)).each do |i|
    @memory[i] = f.sysread(4).unpack("i")[0]
  end
end
# -----------------------------------------------------------------------------
(@cells..1000000).each do |i|
  @memory[i] = 0
end
(0..12).each do |i|
  @inputs[i] = 0
  @ports[i] = 0
end
# -----------------------------------------------------------------------------
ARGV.each do |a|
  if a != "--with" then
    @isp = @isp + 1
    @inputs[@isp] = File.open(a, 'r')
  end
end
# -----------------------------------------------------------------------------
while @ip < 1000000 do
  process()
  @ip = @ip + 1
end

File: vm/complete/retro.py
#!/usr/bin/env python

# Ngaro VM
# Copyright (c) 2010 - 2012, Charles Childers
# Copyright (c) 2011 Greg Copeland ( optimizations and process() rewrite )
# Copyright (c) 2012 Michal J Wallace ( --dump )
# -----------------------------------------------------------------------------

import os, sys, math, time, struct
from struct import pack, unpack

set_termio = False
try:
    import termios
    set_termio = True
except: pass

EXIT = 0x0FFFFFFF # preliminary value

# -----------------------------------------------------------------------------
def rxDivMod( a, b ):
  x = abs(a)
  y = abs(b)
  q, r = divmod(x, y)

  if a < 0 and b < 0:
    r *= -1
  elif a > 0 and b < 0:
    q *= -1
  elif a < 0 and b > 0:
    r *= -1
    q *= -1

  return q, r


def rxGetInput( inputs ):
  a = 0
  if inputs[-1] != 0:
    a = inputs[-1].read(1)
    if len(a) == 0:
      inputs.pop()
      a = 0
    else:
      a = ord(a)
  else:
    a = ord(sys.stdin.read(1))
  return a


# -----------------------------------------------------------------------------
def rxHandleDevices( ip, stack, address, ports, memory, files, inputs ):
  ports[0] = 1
  if ports[1] == 1:
    ports[1] = rxGetInput( inputs )
    if ports[1] == 13:
      ports[1] = 10
  if ports[2] == 1:
    #  TODO: account for unicode correctly
    if stack[-1] > 0 and stack[-1] < 128:
      if stack[-1] == 8:
        sys.stdout.write(chr(stack.pop()))
        sys.stdout.write(chr(32))
        sys.stdout.write(chr(8))
      else:
        sys.stdout.write(chr(stack.pop()))
    else:
      sys.stdout.write("\033[2J\033[1;1H")
      stack.pop()
    sys.stdout.flush()
    ports[2] = 0
  if ports[4] ==  1:    # file save
    f = open('retroImage', 'wb')
    i = 0
    while i < memory[3]:
      f.write(struct.pack('i', memory[i]))
      i += 1
    f.close()
    ports[4] = 0
  if ports[4] ==  2:    # file include
    name = ''
    i = stack.pop()
    while memory[i] != 0:
      name = name + chr(memory[i])
      i += 1
    f = open(name, 'r')
    inputs.append( f )
    ports[4] = 0
  if ports[4] == -1:    # file open
    slot = 0
    i = 1
    name = ''
    while i < 8:
      if files[i] == 0:
        slot = i
      i += 1
    mode = stack.pop()
    i = stack.pop()
    while memory[i] != 0:
      name = name + chr(memory[i])
      i += 1
    if slot > 0:
      if mode == 0:
        if os.path.exists(name):
          files[slot] = open(name, 'r')
        else:
          slot = 0
      elif mode == 1:
        files[slot] = open(name, 'w')
      elif mode == 2:
        files[slot] = open(name, 'a')
      elif mode == 3:
        if os.path.exists(name):
          files[slot] = open(name, 'r+')
        else:
          slot = 0
    ports[4] = slot;
  elif ports[4] == -2:  # file read
    slot = stack.pop()
    ports[4] = ord(files[slot].read(1))
  elif ports[4] == -3:  # file write
    slot = stack.pop()
    files[slot].write(chr(stack.pop()))
    ports[4] = 1
  elif ports[4] == -4:  # file close
    slot = stack.pop()
    files[slot].close()
    files[slot] = 0
    ports[4] = 0
  elif ports[4] == -5:  # file position
    slot = stack.pop()
    ports[4] = files[slot].tell()
  elif ports[4] == -6:  # file seek
    slot = stack.pop();
    pos  = stack.pop();
    ports[4] = files[slot].seek(pos, 0);
  elif ports[4] == -7:  # file size
    slot = stack.pop()
    at = files[slot].tell()
    files[slot].seek(0, 2) # SEEK_END
    ports[4] = files[slot].tell()
    files[slot].seek(at, 0) # SEEK_SET
  if ports[4] == -8:    # file delete
    name = ''
    i = stack.pop()
    while memory[i] != 0:
      name = name + chr(memory[i])
      i = i + 1
    i = 0
    if os.path.exists(name):
      os.remove(name)
      i = 1
    ports[4] = i
  if ports[5] == -1:  # memory size - zero based index
    ports[5] = 1000000 - 1
  elif ports[5] == -2:  # canvas exists?
    ports[5] = 0
  elif ports[5] == -3:  # canvas width
    ports[5] = 0
  elif ports[5] == -4:  # canvas height
    ports[5] = 0
  elif ports[5] == -5:  # stack depth
    ports[5] = len(stack)
  elif ports[5] == -6:  # address stack depth
    ports[5] = len(address)
  elif ports[5] == -7:  # mouse exists?
    ports[5] = len(stack)
  elif ports[5] == -8:  # time
    ports[5] = int(time.time())
  elif ports[5] == -9:  # exit vm
    ip = EXIT
    ports[5] = 0
  elif ports[5] == -10: # environment variable
    req = stack.pop()
    dest = stack.pop()
    key = ''
    while memory[req] != 0:
      key = key + chr(memory[req])
      req += 1
    req = os.getenv(key, default="/")
    memory[dest] = 0
    i = 0
    while i < len(req):
      memory[dest] = ord(req[i])
      i += 1
      dest += 1
      memory[dest] = 0
    ports[5] = 0
  elif ports[5] == -11: # console width
    ports[5] = 0
  elif ports[5] == -12: # console height
    ports[5] = 0
  elif ports[5] == -13: # bits per cell
    ports[5] = 32
  elif ports[5] == -14: # endian
    ports[5] = 0
  elif ports[5] == -15: # console extensions
    ports[5] = -1
  elif ports[5] == -16: # data stack max depth
    ports[5] = 128
  elif ports[5] == -17: # address stack max depth
    ports[5] = 1024

  if ports[8] == 1:
    ports[8] = 0
    a = stack[-2]
    stack[-2] = stack[-1]
    stack[-1] = a
    sys.stdout.write("\33[" + str(stack.pop()) + ";" + str(stack.pop()) + "H")
  elif ports[8] == 2:
    ports[8] = 0
    sys.stdout.write("\33[3" + str(stack.pop()) + "m")
  elif ports[8] == 3:
    ports[8] = 0
    sys.stdout.write("\33[4" + str(stack.pop()) + "m")
  else:
    ports[8] = 0

  return ip


# -----------------------------------------------------------------------------
def process( memory, inputs ):
  ip = 0
  global EXIT
  EXIT = len( memory )
  stack = [] * 128
  address = [] * 1024
  ports = [0] * 12
  files = [0] * 8

  while ip < EXIT:
    opcode = memory[ip]

    # There are 31 opcodes ( 0 .. 30 ).
    # Instructions above this range are treated as implicit calls.

    if opcode > 30:
      address.append( ip )
      ip = memory[ip] - 1
      try:
        while memory[ip + 1] == 0:
          ip += 1
      except IndexError: pass

    else:

      if   opcode ==  0:   # nop
        pass

      elif opcode ==  1:   # lit
        ip += 1
        stack.append( memory[ip] )

      elif opcode ==  2:   # dup
        stack.append( stack[-1] )

      elif opcode ==  3:   # drop
        stack.pop()

      elif opcode ==  4:   # swap
        a = stack[-2]
        stack[-2] = stack[-1]
        stack[-1] = a

      elif opcode ==  5:   # push
        address.append( stack.pop() )

      elif opcode ==  6:   # pop
        stack.append( address.pop() )

      elif opcode ==  7:   # loop
        stack[-1] -= 1
        if stack[-1] != 0 and stack[-1] > -1:
          ip += 1
          ip = memory[ip] - 1
        else:
          ip += 1
          stack.pop()

      elif opcode ==  8:   # jump
        ip += 1
        ip = memory[ip] - 1
        if memory[ip + 1] == 0:
          ip += 1
          if memory[ip + 1] == 0:
            ip += 1

      elif opcode ==  9:   # return
        ip = address.pop()
        if memory[ip + 1] == 0:
          ip += 1
          if memory[ip + 1] == 0:
            ip += 1

      elif opcode == 10:   # >= jump
        ip += 1
        a = stack.pop()
        b = stack.pop()
        if b > a:
          ip = memory[ip] - 1

      elif opcode == 11:   # <= jump
        ip += 1
        a = stack.pop()
        b = stack.pop()
        if b < a:
          ip = memory[ip] - 1

      elif opcode == 12:   # != jump
        ip += 1
        a = stack.pop()
        b = stack.pop()
        if b != a:
          ip = memory[ip] - 1

      elif opcode == 13:   # == jump
        ip += 1
        a = stack.pop()
        b = stack.pop()
        if b == a:
          ip = memory[ip] - 1

      elif opcode == 14:   # @
        stack[-1] = memory[stack[-1]]

      elif opcode == 15:   # !
        mi = stack.pop()
        memory[mi] = stack.pop()

      elif opcode == 16:   # +
        t = stack.pop()
        stack[ -1 ] += t
        stack[-1] = unpack('=l', pack('=L', stack[-1] & 0xffffffff))[0]

      elif opcode == 17:   # -
        t = stack.pop()
        stack[-1] -= t
        stack[-1] = unpack('=l', pack('=L', stack[-1] & 0xffffffff))[0]

      elif opcode == 18:   # *
        t = stack.pop()
        stack[-1] *= t
        stack[-1] = unpack('=l', pack('=L', stack[-1] & 0xffffffff))[0]

      elif opcode == 19:   # /mod
        a = stack[-1]
        b = stack[-2]
        stack[-1], stack[-2] = rxDivMod( b, a )
        stack[-1] = unpack('=l', pack('=L', stack[-1] & 0xffffffff))[0]
        stack[-2] = unpack('=l', pack('=L', stack[-2] & 0xffffffff))[0]

      elif opcode == 20:   # and
        t = stack.pop()
        stack[-1] &= t

      elif opcode == 21:   # or
        t = stack.pop()
        stack[-1] |= t

      elif opcode == 22:   # xor
        t = stack.pop()
        stack[-1] ^= t

      elif opcode == 23:   # <<
        t = stack.pop()
        stack[-1] <<= t

      elif opcode == 24:   # >>
        t = stack.pop()
        stack[-1] >>= t

      elif opcode == 25:   # 0;
        if stack[-1] == 0:
          stack.pop()
          ip = address.pop()

      elif opcode == 26:   # inc
        stack[-1] += 1

      elif opcode == 27:   # dec
        stack[-1] -= 1

      elif opcode == 28:   # in
        t = stack[-1]
        stack[-1] = ports[t]
        ports[t] = 0

      elif opcode == 29:   # out
        pi = stack.pop()
        ports[ pi ] = stack.pop()

      elif opcode == 30:   # wait
        # Only call if we have pending I/O
        if ports[0] == 0:
          ip = rxHandleDevices( ip, stack, address, ports, memory, files, inputs )

    ip += 1
  return stack, address, memory


def dump( stack, address, memory ):
  """
  dumps the vm state. for use with /test/ngarotest.py
  triggered when --dump passed to the program.
  """
  fsep = chr( 28 ) # file separator
  gsep = chr( 29 ) # group separator
  sys.stdout.write( fsep )
  sys.stdout.write( ' '.join( map( str, stack )))
  sys.stdout.write( gsep )
  sys.stdout.write( ' '.join( map( str, address )))
  sys.stdout.write( gsep )
  sys.stdout.write( ' '.join( map( str, memory )))


# -----------------------------------------------------------------------------
def run():

  imgpath = None
  dump_after = False
  
  # inputs contains the stack of file objects to read input from
  # the 0 just acts as a sentinel to indicate there are no more
  # inputs and retro should go back to reading from the console.
  inputs = [0]

  i = 1 # sys.argv[0] is 'retro.py', okay to skip
  while i < len(sys.argv):
    arg = sys.argv[i]
    if arg == "--with":
      i += 1
      inputs.append( open( sys.argv[i], 'r' ))
    elif arg == "--dump":
      dump_after = True
    elif arg == "--image":
      i += 1
      arg = sys.argv[i]
      if os.path.exists(arg): imgpath = arg
    else:
      print "file not found: %r" % arg
      sys.exit(-2)
    i += 1
  del i
  imgpath = imgpath or 'retroImage'

  # print "imgpath:", imgpath
  cells = int(os.path.getsize(imgpath) / 4)
  f = open( imgpath, 'rb' )
  memory = list(struct.unpack( cells * 'i', f.read() ))
  f.close()

  if not dump_after:
    remaining = 1000000 - cells
    memory.extend( [0] * remaining )

  global set_termio
  if set_termio:
    try: old = termios.tcgetattr(sys.stdin.fileno())
    except: set_termio = False
  if set_termio:
    t = termios.tcgetattr(sys.stdin.fileno())
    t[3] = t[3] & ~(termios.ICANON | termios.ECHO | termios.ISIG)
    t[0] = 0
    termios.tcsetattr(sys.stdin.fileno(), termios.TCSANOW, t)
  try:
    stack, address, memory = process(memory, inputs)
    if dump_after: dump(stack, address, memory)
  except:
    raise
  finally:
    if set_termio:
      termios.tcsetattr(sys.stdin.fileno(), termios.TCSANOW, old)

# -----------------------------------------------------------------------------
if __name__ == "__main__":
  run()
